File: c/easy/sum_of_root_to_leaf_binary_numbers.c
#include <stdio.h>
#include <stdlib.h>

/* Tree node definition */
typedef struct TreeNode {
    int val;
    struct TreeNode* left;
    struct TreeNode* right;
} TreeNode;


/* DFS helper */
int dfs(TreeNode* node, int current_value) {
    if (node == NULL)
        return 0;

    /* Shift bits and append current bit */
    current_value = (current_value << 1) | node->val;

    /* Leaf node */
    if (node->left == NULL && node->right == NULL)
        return current_value;

    return dfs(node->left, current_value)
         + dfs(node->right, current_value);
}


/* Main function */
int sumRootToLeaf(TreeNode* root) {
    return dfs(root, 0);
}

File: rust/easy/sum_of_root_to_leaf_binary_numbers.rs
#[derive(Debug)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Box<TreeNode>>,
    pub right: Option<Box<TreeNode>>,
}

pub struct Solution;

impl Solution {
    pub fn sum_root_to_leaf(root: Option<Box<TreeNode>>) -> i32 {
        fn dfs(node: &Option<Box<TreeNode>>, current_value: i32) -> i32 {
            match node {
                None => 0,
                Some(n) => {
                    // Shift bits and append current bit
                    let current_value = (current_value << 1) | n.val;

                    // Leaf node
                    if n.left.is_none() && n.right.is_none() {
                        return current_value;
                    }

                    dfs(&n.left, current_value)
                        + dfs(&n.right, current_value)
                }
            }
        }

        dfs(&root, 0)
    }
}